external-crates/move/crates/move-compiler/src/typing/core.rs

```diff
@@ -841,6 +899,23 @@ pub fn make_tvar(_context: &mut Context, loc: Loc) -> Type {
     sp(loc, Type_::Var(TVar::next()))
 }
 
+pub fn make_autoref(context: &mut Context, loc: Loc, ty: Type) -> (RefVar, Type) {
+    use Type_::*;
+    let rv = RefVar::next();
+    let ty = match &ty.value {
+        Type_::Ref(true, inner) => {
+            context.subst.insert_ref_var(rv, RefKind::MutRef);
+            sp(loc, AutoRef(rv, Box::new(*inner.clone())))
+        }
+        Type_::Ref(false, inner) => {
+            context.subst.insert_ref_var(rv, RefKind::ImmRef);
+            sp(loc, AutoRef(rv, Box::new(*inner.clone())))
+        }
+        _ => sp(loc, AutoRef(rv, Box::new(ty.clone()))),
+    };
+    (rv, ty)
+}
```

Comment:
I think this is pretty good! But I want to try to force `join` into doing more work. 
Big ideas here:
- Let realize_autoref do the heavy lifting
- Make sure we are handling `Var` cases correctly
- The ref cases should get blown away by ready_tvars

```suggestion
pub fn make_autoref(context: &mut Context, loc: Loc, ty: Type) -> (RefVar, Type) {
    use Type_::*;
    let rv = RefVar::next();
    let kind_opt = determine_autoref_kind(&context.subst, &ty);
    if let Some(kind) = kind_opt {
        context.subst.insert_ref_var(rv, kind);
    }
    // this setup should let realize_autoref do the work
    (rv, sp(loc, AutoRef(rv, Box::new(ty))))
}


fn determine_autoref_kind(subst: &Subst, &ty: Type) -> Option<RefKind> {
    use Type_::*;
    match &ty.value {
        Ref(true, _) => Some(RefKind::MutRef),
        Ref(false, _) => Some(RefKind::ImmRef),
        AutoRef(rv, _) => Some(RefKind::Forwarded(*rv)),
        Var(v) => {
            match subst.get(v) {
                None => None, 
                Some(inner) => determine_autoref_kind(subst, inner)
            }
        }
    }
}


```